src/Workspaces/Core/Portable/Workspace/Workspace_Registration.cs

```diff
@@ -58,12 +55,27 @@ protected void UnregisterText(SourceTextContainer textContainer)
 
             var registration = GetWorkspaceRegistration(textContainer);
 
+            // guard us from being called with wrong text container.
             if (registration.Workspace == this)
             {
-                registration.SetWorkspaceAndRaiseEvents(null);
+                RaiseTextRegistrationChangedEvents(registration, newWorkspace: null);
             }
         }
 
+        private void RaiseTextRegistrationChangedEvents(WorkspaceRegistration registration, Workspace newWorkspace, [CallerMemberName] string eventName = null)
+        {
+            var oldWorkspace = registration.Workspace;
+
+            // set workspace right away, but events are lazy. 
+            // people, who is checking Registration.Workspace directly, will see the change right away but
```

Comment:
who does this though?   Is there a benefit to not just making the entire thing lazy?

Reply:
I am not sure I am following you. this is how whole Roslyn Workspace eventing works.